src/test/scala/mesosphere/marathon/integration/setup/ServiceMock.scala

```diff
@@ -0,0 +1,221 @@
+package mesosphere.marathon.integration.setup
+
+import java.util.Date
+import javax.servlet.http.{ HttpServletRequest, HttpServletResponse }
+
+import mesosphere.marathon.integration.setup.ServiceMock.Plan
+import org.eclipse.jetty.server.handler.AbstractHandler
+import org.eclipse.jetty.server.{ Request, Server }
+import play.api.libs.json.{ Json, Writes }
+import scala.util.parsing.combinator.RegexParsers
+
+/**
+  * ServiceMock that speaks the service upgrade protocol v1 + admin handlers.
+  *
+  * GET  /v1/plan -> returns the plan
+  * POST  /v1/plan/continue -> continues the plan with the command flag (decision point)
+  * POST  /v1/plan/interrupt -> This request changes the “has_decision_point” property of the next block to true.
+  * POST  /v1/plan/restart -> This request changes the “status” property of the indicated block to Pending.
+  * POST  /v1/plan/force_complete -> This request changes the “status” property of the indicated block to Complete.
+  *
+  * POST /admin/rollback -> rollback the plan
+  * POST /admin/finalize -> finalize the plan
+  * POST /admin/next -> proceed with next step (no decision point)
+  * POST /admin/error -> add an error to the plan
+  *
+  * @param plan the plan to execute on
+  */
+class ServiceMock(plan: Plan) extends AbstractHandler {
+
+  def start(port: Int) {
+    val server = new Server(port)
+    server.setHandler(this)
+    server.start()
+    server.join()
+  }
+
+  override def handle(target: String,
+                      baseRequest: Request,
+                      request: HttpServletRequest,
+                      response: HttpServletResponse): Unit = synchronized {
+    def status[T](t: T, code: Int)(implicit writes: Writes[T]): Unit = {
+      response.getWriter.write(Json.prettyPrint(Json.toJson(t)))
+      response.setHeader("Content-Type", "application/json")
+      response.setStatus(code)
+    }
+    def ok[T](t: T)(implicit writes: Writes[T]): Unit = status(t, 200)
+    def error[T](t: T)(implicit writes: Writes[T]): Unit = status(t, 503)
+
+    (request.getMethod, request.getPathInfo) match {
+      case ("GET", "/v1/plan") =>
+        if (plan.isDone) ok(plan) else error(plan)
+      case ("POST", "/v1/plan/continue") =>
+        plan.next(withContinue = true)
+        ok(Json.obj("result" -> s"Received cmd: continue"))
+      case ("POST", "/v1/plan/interrupt") =>
+        plan.allBlocks.dropWhile(_.isInProgress).headOption.foreach(_.decisionPoint = true)
```

Comment:
The `.dropWhile(_.isInProgress)` looks suspicious. If the first block has already been completed, it would not be in progress anymore. Then it would be set as a decision point.

Using `.find` and a condition to signify that the block has not yet started executing seems to make more sense for me.